### ISSUE TITLE
Print Nth field index expression when pressing <Enter>

### ISSUE THREAD

**Issue by klassegeljakt** [junegunn/fzf#1323]
- Category
    - [x ] fzf binary
    - [ ] fzf-tmux script
    - [ ] Key bindings
    - [ ] Completion
    - [ ] Vim
    - [ ] Neovim
    - [ ] Etc.
- OS
    - [ ] Linux
    - [ ] Mac OS X
    - [ ] Windows
    - [ ] Windows Subsystem for Linux
    - [ ] Etc.
- Shell
    - [ ] bash
    - [ ] zsh
    - [ ] fish

It would be nice if there was an option to format the output when pressing <Enter>. For example, I use ripgrep to search for files by their contents. When pressing <Enter> I would like to get the filename rather than the filename:linenumber:contents.

**Comment by OJFord** [comment]
Would you reconsider this issue @junegunn?

Your previous response is also true of `--preview`, we _could_ just do:
```
--preview='echo {} | cut -d: -f1 | cmd -'
```
but it's certainly nicer that we can do:
```
--delimiter=: --preview='cmd {1}'
```
(especially when it gets more complicated, such as more than one field used, or one used in multiple places.)

It just seems weird when scripting `fzf --delimiter=:` already for `preview`/`nth`/`with-nth` to have to pipe to `cut --delimiter=:` if the output needs to be transformed too, IMO.

**Comment by junegunn** [comment]
@OJFord Can you give a more concrete example of what you're trying to accomplish with the new option you're proposing?

**Comment by OJFord** [comment]
@junegunn Sure, for example:
```sh
git log --pretty=oneline --all \
    | fzf --ansi --delimiter=' ' --no-multi --preview='git show --color=always {1}' --with-nth=2.. \
    | cut --delimiter=' ' --field=1
```

Here as you see we're taking the `git log` lines like `<SHA> <Subject>`, using only the `<Subject>` for display, but taking the `<SHA>` as argument to `git show` for the preview.

The result is then also the `<SHA>` (which I use for a range of other commands - I call this one `fzsha`, and use it to wrap others that expect a SHA/ref) - but it's unfortunate IMO that we need to use `cut` for this, necessitating specifying the same `delimiter` again.

**Comment by ajeetdsouza** [comment]
I too think this feature would be incredibly useful. `fzf` is a lot smarter than `cut` when it comes to finding field expressions. Moreover, there is no guarantee that `cut` will produce the same field as `fzf`.

It does seem incomplete that `fzf` supports `--preview` and `--with-nth`, but doesn't support `--output`.

**Comment by pR0Ps** [comment]
Here's another simple case where it helps avoid some calls to `awk`/`cut`:
```bash
# fuzzy killing of processes
fkill() {
    ps -Arc -opid=,user=,pcpu=,rss=,command= | fzf --with-nth 2.. --multi --tiebreak=index --output-nth 1 | xargs kill "$@" --
}
```

I made a couple of proofs of concepts (not PR material) to demonstrate some possible integrations:

1. Provide a template with `--output` (matches `--preview` semantics) (branch: [experimental/output](https://github.com/pR0Ps/fzf/tree/experimental/output) )
```bash
$ echo $'a b c d e f\nx y z' | fzf --multi --output "test {1} {3..}"
test 'x' 'z'
test 'a' 'c d e f'
```

2. Provide an "nth" with `--output-nth` (matches `--nth` and `--with-nth` semantics) (branch [experimental/output-nth](https://github.com/pR0Ps/fzf/tree/experimental/output-nth))
```bash
$ echo $'a b c d e f\nx y z' | fzf --multi --output-nth "1,3.."
x z
a c d e f
```

Personally I'm leaning more towards option 2. It's less complex, less surprising (no quotes are added), and should cover the majority of use cases. Thoughts?

Currently these branches just redefine the `opts.Printer` function to modify the output. This means it ignores `--print0`/`--no-print0` and will also attempt to format the query if `--print-query` is used. Definitely not a good way of doing it.

I'm up for fixing this and doing it right, but is this something that you would merge @junegunn ?

**Comment by alexryndin** [comment]
I'm a newbie in fzf but I was really surprised when I found out that fzf can transform input presentation but cannot produce already (internally) transformed output with something like `--output-nth`.

**Comment by alanlgardner** [comment]
I'm also surprised. I ran into this today while trying to get '~' to show in the fzf cd widget. I could alter the display using the FZF_ALT_C_COMMAND and --with-nth, but I could not change the output to cd. That means the only the only solution is to duplicate the entire fzf cd widget function. (See: https://github.com/junegunn/fzf/issues/1430)

In general, I think another justification for including --output and --output-nth (in addition to the great examples above) is as a hook for users to modify all of the 3rd party tools and libraries that have included it, without having to modify those tools' source code.

**Comment by svenXY** [comment]
another useCase found for this would something along the lines of 

```sh
kubectl get pods -A | fzf --no-sort --header-lines=1 
```

Being able to directly only output e.g. the pod name (== --output-nth=2) would also spare me the hassle of having to cut this out with cut or awk - especially since fzf already knows the fields...

**Comment by natebosch** [comment]
> We can easily post-process the output using cut or awk or whatever, and I think it's good enough.
> 
> ```shell
> echo 'a:b:c' | cut -d: -f1
> ```

I don't think this approach works when combined with `execute()` and some command that produces output. The output from the command ends up going through this pipe too.

If we take the `fshow` function you've shown before, but want to be able to use it in a situation like `git cherry-pick $(fshow)` I don't think it's feasible. Starting with the one you have in https://junegunn.kr/2015/03/browsing-git-commits-with-fzf I changed the binding to `ctrl-o` so that enter can still write a result, and dropped `--graph` so I could use `{1}` for the action instead of grepping for the SHA.

```sh
# Cannot be used to grab a SHA as an argument.
fshow() {
  git log --color=always \
      --format="%C(auto)%h%d %s %C(black)%C(bold)%cr" "$@" |
  fzf --ansi --no-sort --reverse --tiebreak=index --bind=ctrl-s:toggle-sort \
      --bind "ctrl-o:execute(git show --color=always {1})"
}
```

Doesn't work for `git cherry-pick $(fshow)` because of the extra output. It would work if we could add an `--output-nth 1`. The suggested workaround of piping through `cut` breaks the keybind.

```sh
# ctrl-o action is broken, and if it is used the output
# is filled with junk
fshow() {
  git log --color=always \
      --format="%C(auto)%h%d %s %C(black)%C(bold)%cr" "$@" |
  fzf --ansi --no-sort --reverse --tiebreak=index --bind=ctrl-s:toggle-sort \
      --bind "ctrl-o:execute(git show --color=always {1})" | cut -d' ' -f1
}
```

**Comment by junegunn** [comment]
> but want to be able to use it in a situation like git cherry-pick $(fshow)

It's a different problem. `--output-nth 1` will not help in your case.

To grab the final output of fzf, you need to redirect the output of the `git show` process like so:

```sh
--bind "ctrl-o:execute(git show --color=always {1} > /dev/tty)"
```

Also see https://gist.github.com/junegunn/8b572b8d4b5eddd8b85e5f4d40f17236

**Comment by natebosch** [comment]
Redirecting to `/dev/tty` works great. Thanks!

**Comment by OJFord** [comment]
@junegunn For whatever it's worth, I'm still interested in this, I pipe fzf to cut in a number of places repeating field index expressions already used in fzf preview for example for cut.

To expand my example above:
```
git log --pretty=oneline --all \
    | fzf --ansi --delimiter=' ' --no-multi --preview='git show --color=always {1}' --with-nth=2.. \
    | cut --delimiter=' ' --field=1
```

with @pR0Ps suggested `output-nth` it would become simply:
```
git log --pretty=oneline --all \
    | fzf --ansi --delimiter=' ' --no-multi --preview='git show --color=always {1}' --with-nth=2.. --output-nth=1
```

Is that still not something you'd consider supporting? (@pR0Ps did say above he's up for implementing it in the first place if you'd consider the PR. 🙂)

**Comment by junegunn** [comment]
Still no. I'm not convinced that the latter is substantially better than the former. I actually prefer the latter as it follows the [Unix philosophy](https://en.wikipedia.org/wiki/Unix_philosophy) and I think it reads better. Every Unix programmer already knows what the `awk` (or `cut`) command is for, but they may have to look up `man fzf` to learn what `--output-nth` is and how it works.

And note that in that case, you can simply omit `--delimiter` option and use `awk`.

```
       -d, --delimiter=STR
              Field delimiter regex for --nth and --with-nth (default: AWK-style)
```

```sh
git log --pretty=oneline --all \
    | fzf --ansi --no-multi --preview='git show --color=always {1}' --with-nth=2.. \
    | awk '{print $1}'
```

So the benefit of the proposed option here is that it will save two or three keystrokes (`| awk '{print $1}'` vs. `--output-nth 1`), and that doesn't seem like a compelling reason to me.

Let me know if there are real-world examples where this option will really shine.

**Comment by Dkendal** [comment]
For sake of completeness, if one is really motivated to do this you can already with the following binding:

```sh
git worktree list | fzf --bind 'enter:execute(echo {1})+close'
```

**Comment by junegunn** [comment]
Fixed in https://github.com/junegunn/fzf/commit/2b584586ed1caf15429625da981575ee35d407b8.

---

`--bind enter:become(echo {1})` was one way to do it. But if you want to support multiple selections, it gets a little trickier.

```sh
fzf --multi --bind 'enter:become(echo {+1})'
```

* We added `+` flag to pass all the selected items.
    * But, `echo` would print multiple selections on a single line, instead of printing each line on a separate line. You need a different command than `echo`.
    * It doesn't work with a large number of selected items.

**Comment by vimpostor** [comment]
> Fixed in 2b584586ed1caf15429625da981575ee35d407b8

Thanks for adding this!

However, there seems to be a discrepancy between `--with-nth` and `--accept-nth` in that the former includes the delimiter and the latter does not. Is this intended?

E.g. the following will show a selection dialog that shows the line `foo;`, however pressing Enter will print `foo`.

```bash
echo 'foo;bar' | fzf --delimiter ';' --with-nth 1 --accept-nth 1
```

Personally, I think the most reasonable UX would be to strip the delimiter both from the selection dialog and from the output (i.e. make `--with-nth` follow the behaviour of `--accept-nth`).

**Comment by junegunn** [comment]
> Personally, I think the most reasonable UX would be to strip the delimiter both from the selection dialog and from the output

I agree with you, but I've been trying not to break backward compatibility here. But now that both options support "templates", you can work around the problem by passing a template `{1}` instead of `1`.

* https://junegunn.github.io/fzf/releases/0.60.0/#advanced-transformation-using-templates

On the flip side, because the users can replicate the old behavior using the template (e.g. `{1};`, we can consider breaking the backward compatibility.

**Comment by junegunn** [comment]
@vimpostor I created an issue. https://github.com/junegunn/fzf/issues/4274
